services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/CrtErrorHandler.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.transfer.s3.internal;
+
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.exception.SdkServiceException;
+import software.amazon.awssdk.crt.CrtRuntimeException;
+import software.amazon.awssdk.crt.s3.CrtS3RuntimeException;
+import software.amazon.awssdk.utils.StringUtils;
+
+import java.util.Optional;
+
+@SdkInternalApi
+public class CrtErrorHandler {
+
+    /**
+     * This class transform a CRTRunTimeException to Meaningful SdkServiceException.
+     * CrtS3RuntimeException are the exceptions generated due to failures in CRTClient due to S3 Service errors.
+     *
+     * @param crtRuntimeException Exception that is thrown by CRTClient.
+     * @return
+     */
+    public Exception transformException(Exception crtRuntimeException) {
+        Optional<CrtS3RuntimeException> crtS3RuntimeExceptionOptional = getCrtS3RuntimeException(crtRuntimeException);
+
+        return crtS3RuntimeExceptionOptional.isPresent()
+                ? crtS3RuntimeExceptionOptional
+                .filter(item -> StringUtils.isNotBlank(item.getAwsErrorCode())
+                        || StringUtils.isNotBlank(item.getAwsErrorMessage())).map(e -> SdkServiceException.builder()
+                        .statusCode(crtS3RuntimeExceptionOptional.get().getStatusCode())
```

Comment:
Can we create specific S3Exception based on the error code?
For example:
`NoSuchBucket` -> `NoSuchBucketException`. 

That way, customers can handle specific cases without checking the error code by themselves. 

If the error code can't be mapped, I think we should just return a generic `S3ServiceException`

If there's no error code, we should wrap it with `SdkClientException`. 

WDUT?

Reply:
Implemented to get Service side exception.
We might need to update the CrtS3RuntimeException to send requestId so that service exception  is built with this.

Reply:
Can we write down each exception type rather than using reflection? I don't think there are many specific exceptions for get and put. https://github.com/aws/aws-sdk-java-v2/blob/master/services/s3/src/main/resources/codegen-resources/service-2.json#L509 In the future, we will generate the code. I think we should be very careful about using reflection.